pyglmnet/pyglmnet.py

```diff
@@ -26,40 +28,39 @@ def _qu(distr, z, eta):
         qu[z <= eta] = np.exp(z[z <= eta])
     elif distr == 'gaussian':
         qu = z
+        grad_qu = np.ones_like(z)
     elif distr == 'binomial':
         qu = expit(z)
+        grad_qu = expit(z) * (1 - expit(z))
     elif distr == 'probit':
-        qu = probit(z)
-    elif distr == 'gamma':
-        qu = np.log1p(np.exp(z))
-    return qu
+        qu = norm.cdf(z)
+        grad_qu = norm.pdf(z)
+    if return_grad:  # eventually this if conditional should go away
+        return qu, grad_qu
+    else:
+        return qu
 
 
-def _logL(distr, beta0, beta, X, y, eta):
+def _logL(distr, y, y_hat):
     """The log likelihood."""
-    z = beta0 + np.dot(X, beta)
-    l = _qu(distr, z, eta)
-    if distr == 'softplus':
-        logL = np.sum(y * np.log(l) - l)
-    elif distr == 'poisson':
-        logL = np.sum(y * np.log(l) - l)
+    if distr in ['softplus', 'poisson']:
+        logL = np.sum(y * np.log(y_hat) - y_hat)
     elif distr == 'gaussian':
-        logL = -0.5 * np.sum((y - l)**2)
+        logL = -0.5 * np.sum((y - y_hat)**2)
     elif distr == 'binomial':
         # analytical formula
-        # logL = np.sum(y*np.log(l) + (1-y)*np.log(1-l))
+        logL = np.sum(y * np.log(y_hat) + (1 - y) * np.log(1 - y_hat))
 
         # but this prevents underflow
-        z = beta0 + np.dot(X, beta)
-        logL = np.sum(y * z - np.log(1 + np.exp(z)))
+        # z = beta0 + np.dot(X, beta)
+        # logL = np.sum(y * z - np.log(1 + np.exp(z)))
```

Comment:
@pavanramkumar I tried deriving this but somehow I am getting an extra `-z` term. I'll share my derivation with you privately. In any case, I understand that this simplification works only for the canonical link, expit: is that correct? In that case, we should perhaps have an internal check for the link.

In terms of the structure of the code base, I think we should semantically separate 'link' and 'distr'. Right now `_qu(...)` takes 'distr' as an argument but I think it should take 'link'. Then we can have a dictionary which maps the 'distr' to the canonical link.